peepingtom/io_/read/star/reader_functions.py

```diff
@@ -6,28 +6,114 @@
 
 raw data is the dictionary produced by starfile.read()
 """
-from .utils import rln30_df_to_particleblocks, rln31_df_to_particleblocks
+from pathlib import Path
 
+import numpy as np
+import pandas as pd
+import eulerangles
 
-def parse_relion30_3d(raw_data):
+from ....datablocks import ParticleBlock
+
+
+coord_headings = {
+    '3d': [f'rlnCoordinate{axis}' for axis in 'XYZ'],
+    '2d': [f'rlnCoordinate{axis}' for axis in 'XY']
+}
+euler_headings = {
+    '3d': [f'rlnAngle{angle}' for angle in ('Rot', 'Tilt', 'Psi')],
+    '2d': 'rlnAnglePsi'
+}
+shift_headings = {
+    '3d': {
+        '3.0': [f'rlnOrigin{axis}' for axis in 'XYZ'],
+        '3.1': [f'rlnOrigin{axis}Angst' for axis in 'XYZ']
+    },
+    '2d': {
+        '3.0': [f'rlnOrigin{axis}' for axis in 'XY'],
+        '3.1': [f'rlnOrigin{axis}Angst' for axis in 'XYZ']
+    }
+}
+
+pixel_size_headings = ['rlnImagePixelSize']
+micrograph_name_heading = 'rlnMicrographName'
+
+
+def extract_data(df, mode='3.1'):
+    particleblocks = []
+    for micrograph_name, df_volume in df.groupby('rlnMicrographName'):
+        if coord_headings['3d'][-1] in df.columns:
+            dim = '3d'
+        else:
+            dim = '2d'
+
+        name = Path(micrograph_name).stem
+
+        coords = df_volume[coord_headings[dim]].to_numpy(dtype=float)
+        shifts = df_volume.get(shift_headings[dim][mode], pd.Series([0.0])).to_numpy()
+        px_size = df_volume.get(pixel_size_headings, pd.Series([1.0])).to_numpy()
+        shifts = shifts / px_size
+        coords += shifts
+
+        eulers = df_volume.get(euler_headings[dim], pd.Series([0])).to_numpy()
+        if dim == '3d':
+            rotation_matrices = euler2matrix_rln(eulers)
+        else:
+            rotation_matrices = rotangle2matrix(eulers)
```

Comment:
lots of redundant work going on in this loop which is the same for each group, if we ever find things aren't running fast enough this is a candidate for optimisation

Reply:
Yup, I changed back to a more unified code because I was having a really hard time maintaining and debugging the split version from before! I haven't seen any significant performance changes, though.